Tools/CISupport/Shared/de-duper

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+# Copyright (C) 2024 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import os
+import subprocess
+import shutil
+import sys
+
+ROOT = '/var/results'
+STORAGE = 'storage'
+
+
+def parse_args(args):
+    parser = argparse.ArgumentParser(
+        description='Program which de-duplicates PNG files in a directory using hard links'
+    )
+    parser.add_argument(
+        'root', nargs='?',
+        help='Path to root of directory to de-duplicate PNG files in',
+        type=str, default='.',
+    )
+    parser.add_argument(
+        '-s', '--storage',
+        help='Directory to store images the program is linking to',
+        type=str,
+        default='storage',
+    )
+    return parser.parse_args(args)
+
+
+def find(path, filter):
+    process = None
+
+    try:
+        process = subprocess.Popen([
+                'find', path,
+                '-name', filter,
+            ], stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            encoding='utf-8',
+        )
+
+        line = process.stdout.readline()
+        while line:
+            yield line.rstrip()
+            line = process.stdout.readline()
+
+    finally:
+        if process and process.poll() is None:
+            process.kill()
+
+
+def md5(path):
+    result = subprocess.run(
+        ['md5sum', path],
+        encoding='utf-8',
+        capture_output=True
+    )
+    if result.returncode:
+        return None
+    return result.stdout.split()[0]
```

Comment:
Why are we doing this rather than using `hashlib.md5(usedforsecurity=False)`? (Also slightly surprised by the choice of md5; sha1 is typically quicker and what most other tools — not least git — use for content hashing.)

Reply:
Validated that sha1 from Python is faster, using that.